### PR TITLE
Support set auto-self-update mode

### LINKED ISSUE #2576: Setting to only check for a self-update during `rustup update`
**Describe the problem you are trying to solve**
When testing the staging rustup release I had it "upgrade" back to the previous version when I ran `rustup update nightly`. This made me reconsider somehow aliasing in `rustup update` to include `--no-self-update` so that I could control when rustup itself updates.

**Describe the solution you'd like**
Something like `rustup set update-self-update check-only` that will check whether a self-update is available and report it, rather than installing it straight away.

**Notes**

### PR DESCRIPTION
**Pull request by 0xPoe**
close https://github.com/rust-lang/rustup/issues/2576

1. The PR supports three different modes of auto self update: enable/disable/check-only.
 - When enabled, commands such as install or update will update rustup itself. 
 - When disable, commands such as install or update will **not** update rustup itself. Only self udpate will update rustup itself. 
 - When check-only, commands such as install or update will **not** update rustup itself but will check for updates. Only self udpate will update rustup itself. 

2. This PR removes a large number of "--no-self-update" args in tests(setup will turn off auto self update by default.) 
3. Add tests for check-only and enable. 
4. add docs

### CODE REVIEW

**Comment by 0xPoe** [comment]
/hold

The PR cannot be review yet.

- [x] Better code 
- [x] Add tests

File: src/cli/rustup_mode.rs [hunk 1]
```diff
@@ -8,12 +8,13 @@ use std::str::FromStr;
...
 use clap::{App, AppSettings, Arg, ArgGroup, ArgMatches, Shell, SubCommand};
 
+use super::common;
 use super::help::*;
 use super::self_update;
 use super::term2;
 use super::term2::Terminal;
 use super::topical_doc;
-use super::{common, self_update::get_available_rustup_version};
+use crate::cli::self_update::{check_rustup_update, SELF_UPDATE_MODES};
```
Comment (rbtcollins) [inline src/cli/rustup_mode.rs]: Why the change from super::... to crate::cli::... ?

I'm not against it, just curious?

```
use super::{common, self_update::{check_rustup_update, SELF_UPDATE_MODES}}
```

would have worked fine

File: src/cli/rustup_mode.rs [hunk 2]
```diff
@@ -911,31 +923,16 @@ fn check_updates(cfg: &Cfg) -> Result<utils::ExitCode> {
...
-    }
     Ok(utils::ExitCode(0))
 }
 
 fn update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
-    let self_update = !m.is_present("no-self-update") && !self_update::NEVER_SELF_UPDATE;
+    let self_update_mode = cfg.get_self_update_mode()?;
+    let self_update = !m.is_present("no-self-update")
+        && !self_update::NEVER_SELF_UPDATE
+        && self_update_mode == "enable";
```
Comment (rbtcollins) [inline src/cli/rustup_mode.rs]: I am not sure I am interpreting this boolean expression correctly - and even if I am, I'm not certain of the intent or user story here; this new feature needs documenting, and that would be a good place to describe the interactions of all the various bits:
- it might be compiled out
- it might be controlled via the command line
- it might be controlled via state

Do places of settings have precedence, or values of settings, or some mixture of both? Lets write it down clearly and see if it makes sense.

File: src/cli/rustup_mode.rs (hunk 1 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Oops! My IDE imported it for me. Fixed.

File: src/cli/rustup_mode.rs (hunk 2 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Added. Update only if rustup does **not** have the no-self-update feature, and auto-self-update is configured to **enable** and has **no** no-self-update parameter.

All other cases will not be updated.

Priority: no-self-update feature > self_update_mode > no-self-update args.

File: doc/src/basics.md [hunk 3]
```diff
@@ -40,8 +40,14 @@ This is the essence of `rustup`.
 
 ## Keeping rustup up to date
 
-Running `rustup update` also checks for updates to `rustup` and automatically
-installs the latest version. To manually check for updates and install the
+We control the automatic update of rustup by setting the auto-self-update configuration. 
+The auto-self-update setting supports three values: `enable` and `disable` and `check-only`.
```
Comment (kinnison) [inline doc/src/basics.md]:
```suggestion
It is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration
variable.  This setting supports three values: `enable` and `disable` and `check-only`.
```

File: doc/src/basics.md [hunk 4]
```diff
@@ -40,8 +40,14 @@ This is the essence of `rustup`.
...
+
+When it is `enable`, running `rustup update` will also check for updates to `rustup` and automatically install the latest version.
+If it is `disable`, the available updates is not checked and installed. 
+In addition, you can set auto-self-update to `check-only` if you want to check rustup for available updates during `rustup update` as well.
```
Comment (kinnison) [inline doc/src/basics.md]:
```suggestion
* `disable` will ensure that no automatic self updating actions are taken.
* `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.
* `check-only` will cause any automatic self update to check and report on any updates, but not to automatically install them.
```

File: doc/src/basics.md [hunk 5]
```diff
@@ -40,8 +40,14 @@ This is the essence of `rustup`.
...
+
+To manually force check for updates and install the
```
Comment (kinnison) [inline doc/src/basics.md]:
```suggestion
Whether automatic self-update is enabled or not, you can request that Rustup check for, and update it self to the
```

File: doc/src/basics.md [hunk 6]
```diff
@@ -40,8 +40,14 @@ This is the essence of `rustup`.
...
 latest version of `rustup` without updating installed toolchains type `rustup
```
Comment (kinnison) [inline doc/src/basics.md]:
```suggestion
latest version of `rustup` without updating any installed toolchains by running `rustup
```

File: src/cli/rustup_mode.rs [hunk 7]
```diff
@@ -1021,6 +1023,10 @@ fn update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
         cfg.temp_cfg.clean();
     }
 
+    if !self_update && self_update_mode == "check-only" {
```
Comment (kinnison) [inline src/cli/rustup_mode.rs]: Should we check even if `--no-self-update` is specified?  I don't mind if the thought is "yes we should" but I wanted to raise it as a potential inconsistency here.

File: src/config.rs [hunk 8]
```diff
@@ -405,6 +406,22 @@ impl Cfg {
         Ok(())
     }
 
+    pub fn set_auto_self_update(&mut self, mode: &str) -> Result<()> {
```
Comment (kinnison) [inline src/config.rs:409]: I'm not entirely sure I like this use of strings here, I think I'd prefer an enumeration for the mode, but I also won't block this PR on that if you don't want to rework our internal APIs to use an enum.

File: src/notifications.rs [hunk 9]
```diff
@@ -102,6 +104,7 @@ impl<'a> Display for Notification<'a> {
                 name
             ),
             SetProfile(name) => write!(f, "profile set to '{}'", name),
+            SetSelfUpdate(mode) => write!(f, "self update mode to '{}'", mode),
```
Comment (kinnison) [inline src/notifications.rs]:
```suggestion
            SetSelfUpdate(mode) => write!(f, "auto-self-update mode set to '{}'", mode),
```

**Review by kinnison — CHANGES_REQUESTED** [review]
There are a number of whitespace changes in this PR.  I don't mind them too much, but in future it's best if this kind of reformatting is only done in independent commits/PRs.

I have made a number of wording change suggestions.  You do not have to accept them all, but in general I think they're likely to improve matters.

File: src/cli/rustup_mode.rs (hunk 7 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: I think if it's check-only then it's better to always check for updates. I simplified the logic. In fact the first condition is useless.

File: src/config.rs (hunk 8 above)
Comment (0xPoe) [reply src/config.rs:409]: Yes, I totally agree that using enumeration is better. I added a fixme and I will change our code to support enums in the next PR.

File: src/cli/rustup_mode.rs (hunk 7 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Also if the user has set check-only I think they may no longer use --no-self-update.

**Comment by 0xPoe** [comment]
> There are a number of whitespace changes in this PR. I don't mind them too much, but in future it's best if this kind of reformatting is only done in independent commits/PRs.

Got it.

File: src/cli/rustup_mode.rs (hunk 7 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Simply put, --no-self-update only useful when auto_self_update is enable and #[cfg(not(feature = "no-self-update"))].

Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Priority: no-self-update feature > self_update_mode > no-self-update args.

**Comment by 0xPoe** [comment]
@kinnison Could you please take another look? All comments addressed. Thanks!

**Comment by 0xPoe** [comment]
Conflicts have been resolved.

File: doc/src/basics.md [hunk 10]
```diff
@@ -30,20 +30,29 @@ info: downloading self-updates
...
+## Keeping rustup up to date
+
+It is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration
+variable.  This setting supports three values: `enable` and `disable` and `check-only`.
+
+* `disable` will ensure that no automatic self updating actions are taken.
+* `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.
+* `check-only` will cause any automatic self update to check and report on any updates, but not to automatically install them.
+
+Whether automatic self-update is enabled or not, you can request that Rustup check for, and update it self to the
```
Comment (kinnison) [inline doc/src/basics.md]:
```suggestion
Whether automatic self-update is enabled or not, you can request that Rustup check for, and update itself to the
```

File: doc/src/basics.md [hunk 11]
```diff
@@ -30,20 +30,29 @@ info: downloading self-updates
...
 ```
 
-> #### Disable automatic self-updates
-> `rustup` will automatically update itself at the end of any toolchain installation.
-> You can prevent this automatic behaviour by passing the `--no-self-update` argument
-> when running `rustup update` or `rustup toolchain install`.
+**Note**: If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any
+toolchain installation as well.  You can prevent this automatic behaviour by
+passing the `--no-self-update` argument when running `rustup update` or
+`rustup toolchain install`.
```
Comment (kinnison) [inline doc/src/basics.md]: A previous change deliberately called this out in a block quote with its own low-tier heading.  I suggest you don't undo that since it's a good readability improvement.

File: src/cli/rustup_mode.rs [hunk 12]
```diff
@@ -1578,6 +1585,11 @@ fn set_profile(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
     Ok(utils::ExitCode(0))
 }
 
+fn set_auto_self_update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
+    cfg.set_auto_self_update(&m.value_of("auto-self-update-mode").unwrap())?;
+    Ok(utils::ExitCode(0))
+}
+
```
Comment (kinnison) [inline src/cli/rustup_mode.rs]: I wonder if this function should warn the user if rustup was built with the `no-self-update` feature and they set it to `enable` ?

**Review by kinnison — CHANGES_REQUESTED** [review]
A couple of doc tweaks and a question which might be "no need" or might lead to a slight code change - otherwise I think this is nearly ready to merge.

File: doc/src/basics.md (hunk 10 above)
Comment (0xPoe) [reply doc/src/basics.md]: Fixed.

File: doc/src/basics.md (hunk 11 above)
Comment (0xPoe) [reply doc/src/basics.md]: Added.

File: src/cli/rustup_mode.rs (hunk 12 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Added.

**Comment by 0xPoe** [comment]
Commits squashed.

File: doc/src/basics.md [hunk 13]
```diff
@@ -30,20 +30,30 @@ info: downloading self-updates
...
 
-> #### Disable automatic self-updates
-> `rustup` will automatically update itself at the end of any toolchain installation.
-> You can prevent this automatic behaviour by passing the `--no-self-update` argument
-> when running `rustup update` or `rustup toolchain install`.
+### Disable automatic self-updates with the parameter of the command
+If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any
+toolchain installation as well.  You can prevent this automatic behaviour by
+passing the `--no-self-update` argument when running `rustup update` or
+`rustup toolchain install`.
```
Comment (kinnison) [inline doc/src/basics.md]: You still seem to have removed the blockquote (`>`) here which was added to provide callout clarity.

```suggestion
> ### Disable automatic self-updates with the parameter of the command
> If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any
> toolchain installation as well.  You can prevent this automatic behaviour by
> passing the `--no-self-update` argument when running `rustup update` or
> `rustup toolchain install`.
```

**Review by kinnison — CHANGES_REQUESTED** [review]

File: doc/src/basics.md (hunk 13 above)
Comment (0xPoe) [reply doc/src/basics.md]: Sorry I missed it. 
Added. Thanks for your review!

**Review by kinnison — APPROVED** [review]
Assuming CI goes green, I think this is okay.  @rbtcollins ?

File: doc/src/basics.md [hunk 14]
```diff
@@ -30,20 +30,30 @@ info: downloading self-updates
...
 info: checking for self-updates
 info: downloading self-updates
 ```
 
-> #### Disable automatic self-updates
-> `rustup` will automatically update itself at the end of any toolchain installation.
-> You can prevent this automatic behaviour by passing the `--no-self-update` argument
-> when running `rustup update` or `rustup toolchain install`.
+> ### Disable automatic self-updates with the parameter of the command
+> If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any
```
Comment (rbtcollins) [inline doc/src/basics.md]: This should probably mention that the default is `enable`?

File: doc/src/basics.md [hunk 15]
```diff
@@ -30,20 +30,30 @@ info: downloading self-updates
...
+It is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration
+variable.  This setting supports three values: `enable` and `disable` and `check-only`.
+
+* `disable` will ensure that no automatic self updating actions are taken.
+* `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.
+* `check-only` will cause any automatic self update to check and report on any updates, but not to automatically install them.
+
+Whether automatic self-update is enabled or not, you can request that Rustup check for, and update itself to the
+latest version of `rustup` without updating any installed toolchains by running `rustup
+self update`:
```
Comment (rbtcollins) [inline doc/src/basics.md]: I think this section also needs to mention the feature that can be disable that disables this capability entirely.

File: src/cli/rustup_mode.rs [hunk 16]
```diff
@@ -911,31 +925,20 @@ fn check_updates(cfg: &Cfg) -> Result<utils::ExitCode> {
...
-        let _ = t.reset();
-        writeln!(t, " : {}", current_version)?;
-    }
     Ok(utils::ExitCode(0))
 }
 
 fn update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
-    let self_update = !m.is_present("no-self-update") && !self_update::NEVER_SELF_UPDATE;
+    let self_update_mode = cfg.get_self_update_mode()?;
+    // Priority: no-self-update feature > self_update_mode > no-self-update args.
```
Comment (rbtcollins) [inline src/cli/rustup_mode.rs:936]: This should be documented somewhere users can see it; since users need to reason about it.

File: src/settings.rs [hunk 17]
```diff
@@ -84,6 +89,7 @@ impl Default for Settings {
             profile: Some("default".to_owned()),
             overrides: BTreeMap::new(),
             pgp_keys: None,
+            auto_self_update: Some("enable".to_owned()),
```
Comment (rbtcollins) [inline src/settings.rs]: For a typed valued in the core here, I'd -really- rather have validated data not strings: I think it will allow fail fast rather than undefined behaviour.

File: src/cli/self_update.rs [hunk 18]
```diff
@@ -1103,6 +1106,32 @@ pub fn get_available_rustup_version() -> Result<String> {
     Ok(String::from(available_version))
 }
 
+pub fn check_rustup_update() -> Result<()> {
```
Comment (rbtcollins) [inline src/cli/self_update.rs:1153]: Just an idea, what about putting this in self-update:

pseudocode:

```
fn self_update(mode: UpdateMode) -> Result<()> {
   if #[cfg(no_self_update)] {
       return Ok(());
   }  else {
      match mode {
         UpdateMode::Disable => { return Ok(()) },
         UpdateMode::CheckOnly => { ... check logic ... }
         UpdateMode::Enable => { ... update logic ... {
     }
  }
}
```

Then you don't need a separate call for checking self updates and can have all the logic about whether self update are possible or not in just one place.

File: doc/src/basics.md (hunk 14 above)
Comment (0xPoe) [reply doc/src/basics.md]: Mentioned.

File: doc/src/basics.md (hunk 15 above)
Comment (0xPoe) [reply doc/src/basics.md]: It was mentioned at the very beginning.

File: src/cli/rustup_mode.rs (hunk 16 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs:936]: The documentation now mentions them in order of priority.

File: src/settings.rs (hunk 17 above)
Comment (0xPoe) [reply src/settings.rs]: A FIXME has been added and I will change our code to support enum in the next PR, now it is difficult to use enum as a configuration item. https://github.com/rust-lang/rustup/blob/3baa99556fd2c06faaf03527a567bc026d5d0899/src/config.rs#L409

In addition this fn now validates the value entered by the user.

File: src/cli/self_update.rs (hunk 18 above)
Comment (0xPoe) [reply src/cli/self_update.rs:1153]: This fn is also used in `rustup check`, so I think we can keep it.

Comment (rbtcollins) [reply src/cli/self_update.rs:1153]: I deliberately suggested an interface taking a mode - that would allow it to be used from `rustup check` as well

File: src/cli/rustup_mode.rs [hunk 19]
```diff
@@ -1578,6 +1585,14 @@ fn set_profile(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
     Ok(utils::ExitCode(0))
 }
 
+fn set_auto_self_update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
+    if self_update::NEVER_SELF_UPDATE {
+        warn!("Your rustup is built with the no-self-update feature, so setting auto-self-update will not have any effect.");
```
Comment (rbtcollins) [inline src/cli/rustup_mode.rs]: Rather than this conversational style, lets be specific about what rustup is involved:

```/home/foo/.cargo/bin/rustup is built with no-self-update feature: setting auto-self-update will not have any effect.```

File: src/settings.rs (hunk 17 above)
Comment (rbtcollins) [reply src/settings.rs]: I understand. I'm expressing a different opinion to @kinnison : I think converting the core to an enum is very straight forward, and there's no need to wait for another PR to do that - this PR itself sat for a week with no movement because the reviewers were busy (very sorry about that!) - but in that time the enum stage could have happened, since we all seem to be agreed that its needed.

If there is some substantial problem doing the enum work that will delay things, then I'm ok merging-and-improving, but I've looked, and it seems very shallow: have a TryInto implementation from str to the Enum type; use it in the toml parser, use it in the config setter, and you're done.

**Review by rbtcollins — CHANGES_REQUESTED** [review]

File: src/cli/self_update.rs (hunk 18 above)
Comment (0xPoe) [reply src/cli/self_update.rs:1153]: The check in the rustup check does not actually check UpdateMode. So I don't think it's necessary to couple it?

File: src/settings.rs (hunk 17 above)
Comment (0xPoe) [reply src/settings.rs]: I agree that these tasks are straightforward, but I actually want to modify the profile-related code as well. So that part of the change is not related to this pr. It only extends the pr merge cycle(It's been fifteen days.). So I would prefer to change the configuration item in the next PR.

File: src/cli/rustup_mode.rs (hunk 19 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Good idea! Added.

File: src/cli/self_update.rs (hunk 18 above)
Comment (0xPoe) [reply src/cli/self_update.rs:1153]: The logic you describe is actually only used in the update method. The value of UpdateMode is not considered elsewhere.
https://github.com/rust-lang/rustup/blob/330a5b2a2bb60389747e58e82c4646bc0a5bd13d/src/cli/rustup_mode.rs#L933

File: doc/src/basics.md [hunk 20]
```diff
@@ -30,20 +30,31 @@ info: downloading self-updates
 
 ## Keeping `rustup` up to date
 
-Running `rustup update` also checks for updates to `rustup` itself and automatically
-installs the latest version. To manually update `rustup` only,
-without updating installed toolchains, type `rustup self update`:
+If your `rustup` build with the `no-self-update` feature, it will not be able to update itself.
```
Comment (rbtcollins) [inline doc/src/basics.md]: Grammar fix: "was built with"

File: doc/src/basics.md [hunk 21]
```diff
@@ -30,20 +30,31 @@ info: downloading self-updates
...
+
+If your `rustup` build without the `no-self-update` feature,
```
Comment (rbtcollins) [inline doc/src/basics.md]: same grammar fix needed here

File: doc/src/basics.md (hunk 20 above)
Comment (rbtcollins) [reply doc/src/basics.md]: Suggest adding that this is the default and that the rustups available from https://rustup.rs are built this way.

File: doc/src/basics.md [hunk 22]
```diff
@@ -30,20 +30,31 @@ info: downloading self-updates
...
-installs the latest version. To manually update `rustup` only,
-without updating installed toolchains, type `rustup self update`:
+If your `rustup` build with the `no-self-update` feature, it will not be able to update itself.
+
+If your `rustup` build without the `no-self-update` feature, 
+it is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration variable. 
+This setting supports three values: `enable` and `disable` and `check-only`.
+
+* `disable` will ensure that no automatic self updating actions are taken.
+* `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.
```
Comment (rbtcollins) [inline doc/src/basics.md]: check for should not be hypenated here. Just say "commands will also check for, and install, any new version of Rustup.

File: doc/src/basics.md [hunk 23]
```diff
@@ -30,20 +30,31 @@ info: downloading self-updates
...
 $ rustup self update
 info: checking for self-updates
 info: downloading self-updates
 ```
 
-> #### Disable automatic self-updates
-> `rustup` will automatically update itself at the end of any toolchain installation.
-> You can prevent this automatic behaviour by passing the `--no-self-update` argument
-> when running `rustup update` or `rustup toolchain install`.
+> ### Disable automatic self-updates with the parameter of the command
```
Comment (rbtcollins) [inline doc/src/basics.md]: The conditional clause makes this hard to process. Perhaps we start over now things are different. How about
```
### Disabling self updates on a per-invocation basis
> Self updates can also be suppressed on individual invocations of `rustup` by passing the argurment `--no-self-update`  when running `rustup update` or `rustup toolchain install`.
```

File: src/cli/rustup_mode.rs [hunk 24]
```diff
@@ -1578,6 +1585,14 @@ fn set_profile(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
     Ok(utils::ExitCode(0))
 }
 
+fn set_auto_self_update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
+    if self_update::NEVER_SELF_UPDATE {
+        warn!("{} is built with the no-self-update feature: setting auto-self-update will not have any effect.",cfg.rustup_dir.display());
```
Comment (rbtcollins) [inline src/cli/rustup_mode.rs]: I think you need arg0 + os.path.realpath, like from proxy_mode.rs
```
        let mut args = crate::process().args_os();
        let arg0 = args.next().map(PathBuf::from);
```
     rustup_dir is just the directory that the common case install path for the binary is: we want the actual path, which for a snap, is some random directory.

File: src/cli/self_update.rs [hunk 25]
```diff
@@ -86,9 +89,53 @@ pub const NEVER_SELF_UPDATE: bool = true;
...
+    }
+}
+
+impl ToString for SelfUpdateMode {
+    fn to_string(&self) -> String {
+        let modes = Self::modes();
+        match self {
+            SelfUpdateMode::Enable => modes[0].to_string(),
+            SelfUpdateMode::Disable => modes[1].to_string(),
+            SelfUpdateMode::CheckOnly => modes[2].to_string(),
```
Comment (rbtcollins) [inline src/cli/self_update.rs]: I don't think the modes array here adds anything: you have connaisence of value between the ordering of the array and the values in the lookup, vs connaisence of value between an inline string here and a string in FromStr; string interning should optimise the binary size the same either way, and so it becomes what is easier to read: 

```
match self {
   SelfUpdateMode::Enable => "enable",
   SelfUpdateMode::Disable => "disable"
   SelfUpdateMode::CheckOnly => "check-only"
}.into()
```

would be my preference - but its ok if you don't choose to change it; just making a note in passing.

File: src/settings.rs [hunk 26]
```diff
@@ -84,6 +91,7 @@ impl Default for Settings {
             profile: Some("default".to_owned()),
             overrides: BTreeMap::new(),
             pgp_keys: None,
+            auto_self_update: Some(SelfUpdateMode::Enable),
```
Comment (rbtcollins) [inline src/settings.rs]: Either this should default to None, or the type should not have Option<> in it.

If we want to support writing the file back with an unset key unset (which I think is probably good), then we need to keep None as the default value and only cast up to Enable on queries. Otherwise new files will be written with a value in them.

And get_self_update_mode already returns Enable when the value is None....

**Review by rbtcollins — COMMENTED** [review]
looking quite good

**Comment by 0xPoe** [comment]
@rbtcollins All comments addressed. Thanks for your review!

**Comment by 0xPoe** [comment]
@rbtcollins @kinnison  Can we merge it? Are there any other issues?

File: src/cli/rustup_mode.rs [hunk 27]
```diff
@@ -1578,6 +1586,17 @@ fn set_profile(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
     Ok(utils::ExitCode(0))
 }
 
+fn set_auto_self_update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
+    if self_update::NEVER_SELF_UPDATE {
+        let mut args = crate::process().args_os();
+        let arg0 = args.next().map(PathBuf::from);
+        let arg0 = arg0.as_ref().and_then(|a| a.to_str());
+        warn!("{} is built with the no-self-update feature: setting auto-self-update will not have any effect.",arg0.ok_or(CLIError::NoExeFilePath)?);
```
Comment (kinnison) [inline src/cli/rustup_mode.rs]:
```suggestion
        let arg0 = arg0.as_ref().and_then(|a| a.to_str());
        let arg0 = arg0.ok_or(CLIError::NoExeFilePath)?;
        warn!("{} is built with the no-self-update feature: setting auto-self-update will not have any effect.",arg0);
```

(I don't mind at all if you don't take this, but I think it'd look neater, the `warn` line is already very long)

**Review by kinnison — APPROVED** [review]
One tiny suggestion, otherwise I'm good with this.  If @rbtcollins is happy then we can merge IMO.

File: src/cli/rustup_mode.rs (hunk 27 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: Fixed. Thanks for your review.

**Comment by 0xPoe** [comment]
It seems the docker image is broken.

File: doc/src/basics.md [hunk 28]
```diff
@@ -30,20 +30,32 @@ info: downloading self-updates
 
 ## Keeping `rustup` up to date
 
-Running `rustup update` also checks for updates to `rustup` itself and automatically
-installs the latest version. To manually update `rustup` only,
-without updating installed toolchains, type `rustup self update`:
+If your `rustup` was build with the `no-self-update` feature, it will not be able to update itself.
```
Comment (rbtcollins) [inline doc/src/basics.md]: should be past tense 

'was built with'

File: src/cli/rustup_mode.rs [hunk 29]
```diff
@@ -1578,6 +1586,20 @@ fn set_profile(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
...
 }
 
+fn set_auto_self_update(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
+    if self_update::NEVER_SELF_UPDATE {
+        let mut args = crate::process().args_os();
+        let arg0 = args.next().map(PathBuf::from);
+        let arg0 = arg0
+            .as_ref()
+            .and_then(|a| a.to_str())
+            .ok_or(CLIError::NoExeFilePath)?;
```
Comment (rbtcollins) [inline src/cli/rustup_mode.rs]: We already have NoExeName : I don't think is different enough to bother with a new error (and as the error is never caught and never repeated, we can just use anyhow!(...)

**Review by rbtcollins — COMMENTED** [review]
Minor nits - I'll fix these I think

**Comment by 0xPoe** [comment]
@rbtcollins @kinnison Thanks again for your review!
